docs/decisions/0015-expiration-improvements.rst

```diff
@@ -0,0 +1,94 @@
+0015 Expiration and Cancellation Data Improvements
+**************************************************
+
+Status
+======
+Proposed - December 2023
+
+Context
+=======
+There are subtleties around the Assignment cancellation and expiration user
+experience in the enterprise learner-portal frontend that require a more nuanced
+and descriptive approach to the data we persist around these states, and
+how that data is represented in API response payloads for consumption
+by the learner-portal frontend.
+
+How expiration currently works
+------------------------------
+There's currently a management command (run via a cron) that toggles the state
+of Assignments to ``CANCELLED`` under the following conditions:
+
+1. The current date is more than 90 days after the *creation* date of the assignment. We use the creation
+   date as a proxy for the date at which an assignment was allocated, which is problematic under
+   the edge case where an assignment is allocated, cancelled, and re-allocated later.
+2. The current date is greater than the inferred enrollment deadline for the assigned course.
+3. The current date is greater than the expiration date of the subsidy associated with the Assignment's
+   access policy record.
+
+Learners are notified of expiration and cancellation via the learner-portal frontend
+------------------------------------------------------------------------------------
+The learner-portal notifies learners of assignment cancellation and expiration.  These notifications
+are required to be dismissable, which means we need to persist state about dismissal
+for assignments in the backend.
+
+Decision
+========
+
+``EXPIRED`` should be a distinct Assignment State
+-------------------------------------------------
+This will help disambiguate behavior and simplify/remove business logic
+from the learner-portal frontend.  We may wish to also store the expiration reason
+for an assignment (i.e. which of the 3 cases above caused the assignment to become expired).
+
+Persist the allocation, cancellation, and expiration dates
```

Comment:
I'm wondering whether the complexity to manage these persisted dates is worth taking on versus only serializing the `earliest-expiration` field via the `credits-available` API as discussed below. That is, will any consumers of the API need the persisted model fields, or only the serialized `earliest-expiration`?

Though, I suppose persisting the model fields would simplify the serialization of `earliest-expiration`?

Reply:
My interpretation is that we would persist all the lifecycle state change dates in-model (allocation date, cancellation date, and expiration date), but the other 4 dates (subsidy expiration date, enrollment deadline 90 day deadline, and earliest-expiration) would all be serialized dynamically and not persisted in-model.

That said, I do feel like out of all the latter 4 dynamic dates, only the last needs to be serialized.  Thoughts?

Reply:
Yeah, Troy's interpretation is what I was going for, I can clarfiy that better in the doc.  I think only serializing the earliest possible expiration date (which is determined dynamically) makes sense to start.  Can clarify that as well.